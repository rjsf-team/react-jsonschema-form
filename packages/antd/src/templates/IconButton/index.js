import React from "react";

import Button from "antd/lib/button";
import ArrowDownOutlined from "@ant-design/icons/ArrowDownOutlined";
import ArrowUpOutlined from "@ant-design/icons/ArrowUpOutlined";
import DeleteOutlined from "@ant-design/icons/DeleteOutlined";
import PlusCircleOutlined from "@ant-design/icons/PlusCircleOutlined";

export default function IconButton(props) {
  const { iconType = "default", icon, ...otherProps } = props;
  return <Button type={iconType} icon={icon} {...otherProps} />;
}

export function AddButton(props) {
  return (
    <IconButton
      title="Add Item"
      {...props}
      block
      iconType="primary"
      icon={<PlusCircleOutlined />}
    />
  );
}

export function MoveDownButton(props) {
  return (
    <IconButton title="Move down" {...props} icon={<ArrowDownOutlined />} />
  );
}

export function MoveUpButton(props) {
  return <IconButton title="Move up" {...props} icon={<ArrowUpOutlined />} />;
}

export function RemoveButton(props) {
  return (
    <IconButton
      title="Remove"
      {...props}
      danger
      iconType="primary"
      icon={<DeleteOutlined />}
    />
  );
}
