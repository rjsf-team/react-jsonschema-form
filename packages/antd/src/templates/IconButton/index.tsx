import Button, { ButtonProps, ButtonType } from 'antd/lib/button';
import ArrowDownOutlined from '@ant-design/icons/ArrowDownOutlined';
import ArrowUpOutlined from '@ant-design/icons/ArrowUpOutlined';
import CopyOutlined from '@ant-design/icons/CopyOutlined';
import DeleteOutlined from '@ant-design/icons/DeleteOutlined';
import PlusCircleOutlined from '@ant-design/icons/PlusCircleOutlined';
import {
  getUiOptions,
  FormContextType,
  IconButtonProps,
  RJSFSchema,
  StrictRJSFSchema,
  TranslatableString,
} from '@rjsf/utils';

// The `type` for IconButtonProps collides with the `type` for `ButtonProps` so omit it to avoid Typescript issue
export type AntdIconButtonProps<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any
> = Omit<IconButtonProps<T, S, F>, 'type'>;

export default function IconButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: AntdIconButtonProps<T, S, F> & ButtonProps
) {
  const { iconType = 'default', icon, uiSchema, registry, ...otherProps } = props;
  return <Button type={iconType as ButtonType} icon={icon} {...otherProps} />;
}

export function AddButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: AntdIconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <IconButton
      title={translateString(TranslatableString.AddItemButton)}
      {...props}
      block
      iconType='primary'
      icon={<PlusCircleOutlined />}
    />
  );
}

export function CopyButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: AntdIconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return <IconButton title={translateString(TranslatableString.CopyButton)} {...props} icon={<CopyOutlined />} />;
}

export function MoveDownButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: AntdIconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <IconButton title={translateString(TranslatableString.MoveDownButton)} {...props} icon={<ArrowDownOutlined />} />
  );
}

export function MoveUpButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: AntdIconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return <IconButton title={translateString(TranslatableString.MoveUpButton)} {...props} icon={<ArrowUpOutlined />} />;
}

export function RemoveButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: AntdIconButtonProps<T, S, F>
) {
  // The `block` prop is not part of the `IconButtonProps` defined in the template, so get it from the uiSchema instead
  const options = getUiOptions<T, S, F>(props.uiSchema);
  const {
    registry: { translateString },
  } = props;
  return (
    <IconButton
      title={translateString(TranslatableString.RemoveButton)}
      {...props}
      danger
      block={!!options.block}
      iconType='primary'
      icon={<DeleteOutlined />}
    />
  );
}
