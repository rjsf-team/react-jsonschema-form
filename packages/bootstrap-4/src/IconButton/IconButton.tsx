import React from "react";
import Button from "react-bootstrap/Button";
import { IoIosRemove } from "react-icons/io";
import { GrAdd } from "react-icons/gr";
import { AiOutlineArrowUp, AiOutlineArrowDown } from "react-icons/ai";

const mappings: any = {
  remove: <IoIosRemove />,
  plus: <GrAdd />,
  "arrow-up": <AiOutlineArrowUp />,
  "arrow-down": <AiOutlineArrowDown />,
};

type IconButtonProps = {
  icon: string;
  className?: string;
  tabIndex?: number;
  style?: any;
  disabled?: any;
  onClick?: any;
};

const IconButton = (props: IconButtonProps) => {
  const { icon, className, ...otherProps } = props;
  return (
    <Button {...otherProps} variant="light" size="sm">
      {mappings[icon]}
    </Button>
  );
};

export default IconButton;
