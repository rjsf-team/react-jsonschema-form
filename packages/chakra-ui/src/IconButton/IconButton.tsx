import { FormContextType, IconButtonProps, RJSFSchema, StrictRJSFSchema, TranslatableString } from '@rjsf/utils';

import { ArrowUpIcon, ArrowDownIcon, CopyIcon, DeleteIcon } from '@chakra-ui/icons';
import ChakraIconButton from './ChakraIconButton';

export function CopyButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: IconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <ChakraIconButton<T, S, F> title={translateString(TranslatableString.CopyButton)} {...props} icon={<CopyIcon />} />
  );
}

export function MoveDownButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: IconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <ChakraIconButton<T, S, F>
      title={translateString(TranslatableString.MoveDownButton)}
      {...props}
      icon={<ArrowDownIcon />}
    />
  );
}

export function MoveUpButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: IconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <ChakraIconButton<T, S, F>
      title={translateString(TranslatableString.MoveUpButton)}
      {...props}
      icon={<ArrowUpIcon />}
    />
  );
}

export function RemoveButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: IconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <ChakraIconButton<T, S, F>
      title={translateString(TranslatableString.RemoveButton)}
      {...props}
      icon={<DeleteIcon />}
    />
  );
}
