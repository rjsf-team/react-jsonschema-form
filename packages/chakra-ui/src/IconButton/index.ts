export { default } from './ChakraIconButton';
export * from './IconButton';
