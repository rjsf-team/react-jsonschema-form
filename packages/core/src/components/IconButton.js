import React from "react";

export default function IconButton(props) {
  const { type = "default", icon, className, ...otherProps } = props;
  return (
    <button
      type="button"
      className={`btn btn-${type} ${className}`}
      {...otherProps}
    >
      <i className={`glyphicon glyphicon-${icon}`} />
    </button>
  );
}
