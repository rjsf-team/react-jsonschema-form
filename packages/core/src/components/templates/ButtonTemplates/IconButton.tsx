import React from "react";
import { IconButtonProps } from "@rjsf/utils";

export default function IconButton(props: IconButtonProps) {
  const { iconType = "default", icon, className, ...otherProps } = props;
  return (
    <button
      type="button"
      className={`btn btn-${iconType} ${className}`}
      {...otherProps}
    >
      <i className={`glyphicon glyphicon-${icon}`} />
    </button>
  );
}

export function MoveDownButton(props: IconButtonProps) {
  return (
    <IconButton
      title="Move down"
      className="array-item-move-down"
      {...props}
      icon="arrow-down"
    />
  );
}

export function MoveUpButton(props: IconButtonProps) {
  return (
    <IconButton
      title="Move up"
      className="array-item-move-up"
      {...props}
      icon="arrow-up"
    />
  );
}

export function RemoveButton(props: IconButtonProps) {
  return (
    <IconButton
      title="Remove"
      className="array-item-remove"
      {...props}
      iconType="danger"
      icon="remove"
    />
  );
}
