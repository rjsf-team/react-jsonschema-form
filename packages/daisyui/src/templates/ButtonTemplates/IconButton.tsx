import { FormContextType, IconButtonProps, RJSFSchema, StrictRJSFSchema, TranslatableString } from '@rjsf/utils';
import { faCopy, faArrowDown, faArrowUp, faTrash } from '@fortawesome/free-solid-svg-icons';
import type { IconDefinition } from '@fortawesome/fontawesome-svg-core';

import DaisyUIButton from './DaisyUIButton';

export function CopyButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: IconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <DaisyUIButton {...props} title={translateString(TranslatableString.CopyButton)} icon={faCopy as IconDefinition} />
  );
}

export function MoveDownButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: IconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <DaisyUIButton
      {...props}
      title={translateString(TranslatableString.MoveDownButton)}
      icon={faArrowDown as IconDefinition}
    />
  );
}

export function MoveUpButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: IconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <DaisyUIButton
      {...props}
      title={translateString(TranslatableString.MoveUpButton)}
      icon={faArrowUp as IconDefinition}
    />
  );
}

export function RemoveButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: IconButtonProps<T, S, F>
) {
  const {
    registry: { translateString },
  } = props;
  return (
    <DaisyUIButton
      {...props}
      title={translateString(TranslatableString.RemoveButton)}
      iconType='danger'
      icon={faTrash as IconDefinition}
    />
  );
}
