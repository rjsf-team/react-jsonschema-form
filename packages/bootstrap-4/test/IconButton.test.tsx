import React from "react";
import IconButton from "../src/IconButton";
import renderer from "react-test-renderer";

describe("IconButton", () => {
  test("simple", () => {
    const tree = renderer
      .create(
        <IconButton
          icon="plus"
          className="someClass"
          onClick={() => void 0}
          disabled={false}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("disable", () => {
    const tree = renderer
      .create(<IconButton icon="remove" disabled />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
