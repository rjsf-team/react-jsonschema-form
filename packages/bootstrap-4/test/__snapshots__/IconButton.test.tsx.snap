// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton disable 1`] = `
<button
  className="btn btn-light btn-sm"
  disabled={true}
  type="button"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M368.5 240h-225c-8.8 0-16 7.2-16 16 0 4.4 1.8 8.4 4.7 11.3 2.9 2.9 6.9 4.7 11.3 4.7h225c8.8 0 16-7.2 16-16s-7.2-16-16-16z"
    />
  </svg>
</button>
`;

exports[`IconButton simple 1`] = `
<button
  className="btn btn-light btn-sm"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12,22 L12,2 M2,12 L22,12"
      fill="none"
      stroke="#000"
      strokeWidth="2"
    />
  </svg>
</button>
`;
